File: src/main/java/dev/qixils/quasicord/db/DatabaseManager.java
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package dev.qixils.quasicord.db;

import com.mongodb.ConnectionString;
import com.mongodb.MongoClientSettings;
import com.mongodb.client.model.Filters;
import com.mongodb.reactivestreams.client.MongoClient;
import com.mongodb.reactivestreams.client.MongoClients;
import com.mongodb.reactivestreams.client.MongoCollection;
import com.mongodb.reactivestreams.client.MongoDatabase;
import dev.qixils.quasicord.Environment;
import dev.qixils.quasicord.Quasicord;
import org.bson.codecs.configuration.CodecRegistries;
import org.bson.codecs.configuration.CodecRegistry;
import org.bson.codecs.pojo.PojoCodecProvider;
import org.bson.conversions.Bson;
import org.checkerframework.checker.nullness.qual.NonNull;
import org.jetbrains.annotations.NotNull;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.io.Closeable;
import java.util.Locale;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Objects;

/**
 * Manages the {@link Quasicord Bot}'s database.
 */
public class DatabaseManager implements Closeable {
	private final @NotNull CodecRegistry pojoRegistry = CodecRegistries.fromRegistries(MongoClientSettings.getDefaultCodecRegistry(),
			CodecRegistries.fromProviders(PojoCodecProvider.builder().automatic(true).build()));
	private final @NotNull MongoClientSettings mongoClientSettings = MongoClientSettings.builder()
			.applyConnectionString(new ConnectionString("mongodb://localhost:27017"))
			.codecRegistry(pojoRegistry)
			.build();
	private final @NotNull MongoClient mongoClient = MongoClients.create(mongoClientSettings);
	protected final @NotNull MongoDatabase database;

	public DatabaseManager(String dbPrefix, Environment environment) {
		this(dbPrefix, Objects.requireNonNull(environment, "environment cannot be null").name().toLowerCase(Locale.ROOT));
	}

	public DatabaseManager(@NotNull String dbPrefix, @NotNull String dbSuffix) {
		Objects.requireNonNull(dbPrefix, "dbPrefix cannot be null");
		Objects.requireNonNull(dbSuffix, "dbSuffix cannot be null");
		this.database = mongoClient.getDatabase(dbPrefix + "-" + dbSuffix);
	}

	// collectionNameOf

	private static String collectionNameOf(Object object) {
		return collectionNameOf(object.getClass());
	}

	private static String collectionNameOf(Class<?> clazz) {
		String collectionName;
		if (clazz.isAnnotationPresent(CollectionName.class))
			collectionName = clazz.getAnnotation(CollectionName.class).value();
		else
			collectionName = clazz.getSimpleName();
		return collectionName.substring(0, Math.min(collectionName.length(), 127));
	}

	// collection

	public <T> @NotNull MongoCollection<T> collection(Class<T> tClass) {
		return collection(collectionNameOf(tClass), tClass);
	}

	public <T> @NotNull MongoCollection<T> collection(String collectionName, Class<T> tClass) {
		return database.getCollection(collectionName, tClass);
	}

	// getAll

	public <T> @NotNull Flux<T> getAll(Class<T> tClass) {
		return Flux.from(collection(tClass).find());
	}

	// getAllBy

	public <T> @NotNull Flux<T> getAllBy(Bson filters, Class<T> tClass) {
		return Flux.from(collection(tClass).find(filters));
	}

	public <T> @NotNull Flux<T> getAllByEquals(Map<String, ?> keyValueMap, Class<T> tClass) {
		Bson filter = Filters.empty();
		for (Entry<String, ?> entry : keyValueMap.entrySet())
			filter = Filters.and(filter, Filters.eq(entry.getKey(), entry.getValue()));
		return getAllBy(filter, tClass);
	}

	// getById

	public <T> @NonNull Mono<T> getById(Object id, Class<T> tClass) {
		return Mono.from(collection(tClass).find(Filters.eq("_id", id)));
	}

	// misc

	public void close() {
		mongoClient.close();
	}
}

File: src/main/java/dev/qixils/quasicord/db/DatabaseManager.kt
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package dev.qixils.quasicord.db

import com.mongodb.ConnectionString
import com.mongodb.MongoClientSettings
import com.mongodb.kotlin.client.coroutine.MongoClient
import com.mongodb.kotlin.client.coroutine.MongoCollection
import com.mongodb.kotlin.client.coroutine.MongoDatabase
import dev.qixils.quasicord.Environment
import dev.qixils.quasicord.db.codecs.LocaleCodec
import dev.qixils.quasicord.db.codecs.ZoneIdCodec
import org.bson.codecs.configuration.CodecRegistries
import org.bson.codecs.configuration.CodecRegistry
import org.bson.codecs.pojo.PojoCodecProvider
import java.io.Closeable

/**
 * Manages the [Bot][Quasicord]'s database.
 */
class DatabaseManager(dbPrefix: String, dbSuffix: String) : Closeable {
    private val pojoRegistry: CodecRegistry = CodecRegistries.fromRegistries(
		CodecRegistries.fromCodecs(ZoneIdCodec, LocaleCodec),
        MongoClientSettings.getDefaultCodecRegistry(),
        CodecRegistries.fromProviders(PojoCodecProvider.builder().automatic(true).build())
    )
    private val mongoClientSettings: MongoClientSettings = MongoClientSettings.builder()
        .applyConnectionString(ConnectionString("mongodb://localhost:27017")) // TODO: env
        .codecRegistry(pojoRegistry)
        .build()
    private val mongoClient: MongoClient = MongoClient.create(mongoClientSettings)
    protected val database: MongoDatabase = mongoClient.getDatabase("$dbPrefix-$dbSuffix")
	private val caches = mutableMapOf<Class<*>, CachedQueries<*>>()

	constructor(dbPrefix: String, environment: Environment) : this(
        dbPrefix,
        environment.name.lowercase(),
    )

	// collection
    fun <T : Any> collection(tClass: Class<T>): MongoCollection<T> {
        return collection(collectionNameOf(tClass), tClass)
    }

	inline fun <reified T : Any> collection(): MongoCollection<T> {
		return collection(T::class.java)
	}

    fun <T : Any> collection(collectionName: String, tClass: Class<T>): MongoCollection<T> {
		// TODO: indices
        return database.getCollection(collectionName, tClass)
    }

	inline fun <reified T : Any> collection(collectionName: String): MongoCollection<T> {
		return collection(collectionName, T::class.java)
	}

	fun <T : Any> cache(tClass: Class<T>): CachedQueries<T> {
		return caches.computeIfAbsent(tClass) { CachedQueries(collection(tClass)) } as CachedQueries<T>
    }

	inline fun <reified T : Any> cache(): CachedQueries<T> {
		return cache(T::class.java)
	}

    // misc
    override fun close() {
        mongoClient.close()
    }

    companion object {
        // collectionNameOf
        private fun collectionNameOf(obj: Any): String {
            return collectionNameOf(obj.javaClass)
        }

        private fun collectionNameOf(clazz: Class<*>): String {
			val collectionName =
				if (clazz.isAnnotationPresent(CollectionName::class.java))
					clazz.getAnnotation(CollectionName::class.java).value
				else
					clazz.getSimpleName()
            return collectionName.substring(0, collectionName.length.coerceAtMost(127))
        }
    }
}
